docs/designs/restore-and-gc.md

```diff
@@ -0,0 +1,217 @@
+# Restore
+
+Docfx `restore` helps you to restore all you dependency repositories/files(defined in [config](config.md), fallback source repository and theme repository) into your local `%DOCFX_APPDATA_PATH%` for build preparation.
+
+The default `%DOCFX_APPDATA_PATH%` is `%USERPROFILE%/.docfx`, but you can reset it to any other place.
+
+The restore need make sure:
+
+  - Once one version of file/repository is restored, this version of the file/repository should **NOT** be updated until be cleaned up by [GC](#Garbage collection). To make sure the `build`s which using this version are not broken causing by another `restore`.
+  - Provide a mechanism to lock down the version of dependency file/repository for different `build`s, just like [npm package locks](https://docs.npmjs.com/files/package-locks)
+
+## Restore files
+
+There are some cases that the docfx `build` depends on external files for help. 
+
+For example, if you want the docfx to resolve `gitHub contributors` from building file git commit history, the [GitHub User Cache](github-user-cache.md) file can be set to reduce the API calls to [GitHub](https://github.com). The file can be just a http(s) url, which `restore` will help to download for `build` usage.
+
+```yml
+github:
+  userCache: https:///some-blob-service/github-user-cache.json
+```
+
+The restored files will be stored at `%DOCFX_APPDATA_PATH%/downloads/{url-short-path}/{version}+{etag}` like:
+
+```text
+`%DOCFX_APPDATA_PATH%`
+    | - downloads
+        | - raw.gith..tent.com+docascode+docfx-te..ndencies+62b0448+extend1.yml+dc363b0e
+            | - 18b265bc4e9a6b4fdff8b5210caf9e3541f4cb2d+%22f02fde5b4f3817266458170ba048c7fc2de46287%22
+            | - 3111eb522b092c9cef5cd0ccd86892cf847592d6+%2299a5dbb8d73da1ec452d74c0ab45b11ebe8615a0%22
+
+```
+
+- `{url-short-path}` is calculated from `file` url
```

Comment:
>is calculated from `file` url [](start = 21, length = 29)

what is the calculate logic? should we add it here?

Reply:
the above `restore file` and `restore git`